### DISCUSSION TITLE
Pull request workflow

### DISCUSSION THREAD

**Discussion by pokey** [cursorless-dev/cursorless#1752] [Q&A]
I've been using a pull request workflow where I force-push with impunity before a PR has been reviewed, but then stop rewriting history once it has had its first review. This compromise lets me avoid tons of commits during development, but makes it easier for reviewers to see what's changed since they last reviewed, and makes it easier if they want to make some tweaks to your branch. Github and the VSCode github extension break pretty badly if you force push

**Comment by josharian** [comment]
What I think is that I miss Gerrit. :)

Anyway, if that’s the preferred workflow, I can roll with it. Always happy to do as the Romans do.

Do you let github do the final rebase down to a single commit at merge time? Does that mean that I need to go to the website and keep the PR text up-to-date, because that is what will be the ultimate commit message?

It seems that I should probably get a brain dump from you about git branchless, because it seems that that is probably the closest way to approximate my preferred workflow using GitHub. I installed locally and poked at it a little, but a quick primer would help a lot.

**Reply by auscompgeek** [reply]
> Do you let github do the final rebase down to a single commit at merge time?

Yep, all PRs are squash merged. GitHub's merge queue forces us to use only one of the three merge methods, and we had always been using squash merge.

**Reply by pokey** [reply]
and yeah we use PR description as ultimate commit message. I def would encourage you to keep the PR description up to date rather than commit message, as I completely ignore the latter 😅

**Comment by auscompgeek** [comment]
Graphite lets you compare between force pushes, I believe. Their pricing feels a bit aggressive but just using the review dashboard we'd probably fit fairly comfortably within their free tier limits.

**Reply by pokey** [reply]
Have you messed with graphite much? I really struggled to get their cli to work for my workflow

**Reply by auscompgeek** [reply]
Since they've pushed all their large customers to pay, I haven't been able to use their CLI at all 😅

My work would've only gotten value out of the CLI if we paid; their other bells and whistles we didn't find useful given our workflow constraints. The review dashboard is still useful without paying though.
